temporian/utils/rtcheck.py

```diff
@@ -0,0 +1,340 @@
+        num_checks = min(len(value), _NUM_CHECK_STRUCT)
+        idx = 0
+        for v in value:
+            if idx >= num_checks:
+                break
+            idx += 1
+
+            _check_annotation(
+                trace.add_context(
+                    "When checking the content of a list, set or tuple."
+                ),
+                is_compiled,
+                v,
+                annotation_args[0],
+            )
+    else:
+        trace.exception(
+            "List and set annotations require zero or one arguments."
+        )
+
+
+def _check_annotation_union(
+    trace: _Trace, is_compiled: bool, value, annotation_args
+):
+    for arg in annotation_args:
+        match = True
+        try:
+            _check_annotation(
+                trace.add_context("When checking the content of a union."),
+                is_compiled,
+                value,
+                arg,
+            )
+        except ValueError:
+            match = False
+        if match:
+            return
+
+    trace.exception(f"Cannot match any item of the union {annotation_args}.")
+
+
+def _check_annotation_tuple(
+    trace: _Trace, is_compiled: bool, value, annotation_args
+):
+    if len(annotation_args) != len(value):
+        trace.add_context("Wrong number of items in tuple")
+
+    for val, arg in zip(value, annotation_args):
+        _check_annotation(
+            trace.add_context("When checking the content of a tuple."),
+            is_compiled,
+            val,
+            arg,
+        )
+
+
+def _check_annotation_dict(
+    trace: _Trace, is_compiled: bool, value, annotation_args
+):
+    if len(annotation_args) == 0:
+        pass  # No sub-type to check
+    elif len(annotation_args) == 2:
+        num_checks = min(len(value), _NUM_CHECK_STRUCT)
+        idx = 0
+        for k, v in value.items():
+            if idx >= num_checks:
+                break
+            idx += 1
+
+            _check_annotation(
+                trace.add_context("When checking a dict key."),
+                is_compiled,
+                k,
+                annotation_args[0],
+            )
+            _check_annotation(
+                trace.add_context("When checking a dict value."),
+                is_compiled,
+                v,
+                annotation_args[1],
+            )
+
+    else:
+        trace.exception("Dict annotation require zero or two arguments.")
+
+
+def rtcheck(fn):
+    """Checks the input arguments and output value of a function.
+
+    Usage example:
+
+        @rtcheck
+        def f(a, b: int, c: str = "aze") -> List[str]:
+            del a
+            del b
+            del c
+            return ["a", "b"]
+
+        f(1, 2, "a") # Ok
+        f(1, 2, 3) # Fails
+
+    If combined with @compile, @rtcheck should be put before @compile.
+
+    This code only support what is required by Temporian API.
+
+    Does not typing.GenericTypeAlias e.g. list[int]. Use List[int] instead.
```

Comment:
This part is not enough for the full GenericTypeAlias support.

This part is needed because typing get something casted (e.g. List -> list), 


*Fix wording.